### ISSUE TITLE
How to change optimizer and lr scheduler in the middle of training?

### ISSUE THREAD

**Issue by abecciu** [PyTorchLightning/pytorch-lightning#3095]
#### What is your question?

I need to train a model with a pre-trained backbone. For the first 10 epochs, I want to have the backbone completely frozen (ie. not touched by the optimizer). After epoch 10, I want to start training certain layers of the backbone. In regular pytorch, I would instantiate a new optimizer adding the backbone params that I want to train. Then I'd swap both optimizer and lr_scheduler. 

What's the recommended way to do something like this in PL?

**Comment by rohitgr7** [comment]
you can try the `on_train_epoch_start` method in the callback and reconfigure the optimizers and schedulers the way you want. Here are some links that can help:

- https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#on-train-epoch-start
- https://github.com/PyTorchLightning/pytorch-lightning/blob/83ab3eadb6bd32a761fd2e710a7a28778efa0360/pytorch_lightning/accelerators/gpu_backend.py#L39-L48

set:
```python
class SomeCallback(Callback):
    def on_train_epoch_start(self, trainer. pl_module):
        if trainer.current_epoch == 10:
            trainer.optimizers = [new_optimizers]
            trainer.lr_schedulers = trainer.configure_schedulers([new_schedulers])
            trainer.optimizer_frequencies = [] # or optimizers frequencies if you have any

trainer = Trainer(callbacks=[SomeCallback()], ...)
```
I think this should work.

**Comment by abecciu** [comment]
Thanks so much! This looks very promising, I couldn't tell from the docs that the `on_train_epoch_start` receives `trainer` and `pl_module` as params. Seems like a great solution.

I'll try this out and report back.

**Comment by rohitgr7** [comment]
fixed the link

**Comment by abecciu** [comment]
This solution worked great for me. It's important to note that the `on_train_epoch_start` does not exist in older versions of PL.

Thanks again @rohitgr7 for the help!

**Comment by icedpanda** [comment]
Hi @rohitgr7, how to change optimizer and LR scheduler if I am using `ReduceLROnPlateau`?  I am not sure how to set monitor, interval, etc while using `ReduceLROnPlateau`

This is my sample `configure_optimizers` for init :

```python
    def configure_optimizers(self):
        optimizer = optim.Adam(
            filter(
                lambda p: p.requires_grad,
                self.net.parameters()),
            lr=self.lr,
            weight_decay=self.weight_decay)
        scheduler = optim.lr_scheduler.ReduceLROnPlateau(
            optimizer,
            mode="min",
            factor=0.1,
            patience=self.scheduler_patience,
            cooldown=3,
        )
        return {
            "optimizer": optimizer,
            "lr_scheduler": {
                "scheduler": scheduler,
                "monitor": "val/loss",
                "interval": "epoch",
                "frequency": 1,
            },
        }
```

However, how to set this monitor, interval while using on_train_start in the middle of training? 

```python
    def on_train_start(self, trainer, pl_module) -> None:

        if trainer.current_epoch == self._unfreeze_at_epoch:
            print("unfreeze and add param group...")
            pl_module.net.encoder.freeze_backbone(False)
            new_optimizer = optim.Adam(
                filter(
                    lambda p: p.requires_grad,
                    pl_module.net.encoder.parameters()),
                lr=pl_module.lr,
                weight_decay=pl_module.weight_decay)
            new_schedulers = optim.lr_scheduler.ReduceLROnPlateau(
                new_optimizer,
                mode="min",
                factor=0.1,
                patience=pl_module.scheduler_patience,
                cooldown=3,
            )
            trainer.optimizers = [new_optimizer]
            trainer.lr_schedulers = [new_schedulers]
            trainer.optimizer_frequencies =1
```

Thanks in advance

**Comment by lxxue** [comment]
> Hi @rohitgr7, how to change optimizer and LR scheduler if I am using `ReduceLROnPlateau`? I am not sure how to set monitor, interval, etc while using `ReduceLROnPlateau`
> 
> This is my sample `configure_optimizers` for init :
> 
> ```python
>     def configure_optimizers(self):
>         optimizer = optim.Adam(
>             filter(
>                 lambda p: p.requires_grad,
>                 self.net.parameters()),
>             lr=self.lr,
>             weight_decay=self.weight_decay)
>         scheduler = optim.lr_scheduler.ReduceLROnPlateau(
>             optimizer,
>             mode="min",
>             factor=0.1,
>             patience=self.scheduler_patience,
>             cooldown=3,
>         )
>         return {
>             "optimizer": optimizer,
>             "lr_scheduler": {
>                 "scheduler": scheduler,
>                 "monitor": "val/loss",
>                 "interval": "epoch",
>                 "frequency": 1,
>             },
>         }
> ```
> 
> However, how to set this monitor, interval while using on_train_start in the middle of training?
> 
> ```python
>     def on_train_start(self, trainer, pl_module) -> None:
> 
>         if trainer.current_epoch == self._unfreeze_at_epoch:
>             print("unfreeze and add param group...")
>             pl_module.net.freeze_backbone(False)
>             new_optimizer = optim.Adam(
>                 filter(
>                     lambda p: p.requires_grad,
>                     pl_module.net.parameters()),
>                 lr=pl_module.lr,
>                 weight_decay=pl_module.weight_decay)
>             new_schedulers = optim.lr_scheduler.ReduceLROnPlateau(
>                 new_optimizer,
>                 mode="min",
>                 factor=0.1,
>                 patience=pl_module.scheduler_patience,
>                 cooldown=3,
>             )
>             trainer.optimizers = [new_optimizer]
>             trainer.lr_schedulers = [new_schedulers]
>             trainer.optimizer_frequencies =1
> ```
> 
> Thanks in advance

I just tried the callback solution and found that you cannot set 
```
trainer.lr_schedulers = [new_schedulers]
```
but need to do this
```
trainer.lr_schedulers = trainer._configure_schedulers(
                    schedulers, monitor=None, is_manual_optimization=False)
```
to properly set the schedulers with monitor and add other default configurations

**Comment by wangm23456** [comment]
> What should I do in 2.0.2 ? Why we can't have a function like configure_optimizers()?在2.0.2中我应该做什么？为什么我们不能有一个像configure_optimizers（）这样的函数？

```python
    def setup_optimizers(self, trainer: "pl.Trainer") -> None:
        """Creates optimizers and schedulers.

        Args:
            trainer: the Trainer, these optimizers should be connected to
        """
        if trainer.state.fn != TrainerFn.FITTING:
            return
        assert self.lightning_module is not None
        self.optimizers, self.lr_scheduler_configs = _init_optimizers_and_lr_schedulers(self.lightning_module)

    def setup(self, trainer: "pl.Trainer") -> None:
        """Setup plugins for the trainer fit and creates optimizers.

        Args:
            trainer: the trainer instance
        """
        assert self.accelerator is not None
        self.accelerator.setup(trainer)
        self.setup_optimizers(trainer)
        self.setup_precision_plugin()
        _optimizers_to_device(self.optimizers, self.root_device)
```

So, just self.trainer.strategy.setup(self.trainer)?

**Comment by BrunoBelucci** [comment]
You could use only the `setup_optimizers` method in the strategy to not mess up with the other configurations, so I would say maybe `trainer.strategy.setup_optimizers(trainer)`
